File: tests/functional/enclave/src/end_to_end/builtin_gbdt_train.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use super::*;
use futures::FutureExt;
use teaclave_crypto::TeaclaveFile128Key;
use teaclave_rpc::CredentialService;
use teaclave_test_utils::async_test_case;
#[async_test_case]
pub async fn test_gbdt_training_task() {
    let mut client = authorized_frontend_client().await;
    let function_id = register_gbdt_function(&mut client).await;
    let training_data_id = register_input_file(&mut client).await;

    let crypto = TeaclaveFile128Key::random();
    let output_model_id = register_output_file(&mut client, crypto).await;

    let task_id = create_gbdt_training_task(&mut client, &function_id).await;
    assign_data_to_task(&mut client, &task_id, training_data_id, output_model_id).await;
    approve_task(&mut client, &task_id).await.unwrap();
    invoke_task(&mut client, &task_id).await.unwrap();

    let ret_val = get_task_until(&mut client, &task_id, TaskStatus::Finished).await;
    assert_eq!(&ret_val, "Trained 120 lines of data.");
}

// Authenticate user before talking to frontend service
async fn authorized_frontend_client() -> TeaclaveFrontendClient<CredentialService> {
    let mut api_client = create_authentication_api_client(shared_enclave_info(), AUTH_SERVICE_ADDR)
        .await
        .unwrap();
    let cred = login(&mut api_client, USERNAME, TEST_PASSWORD)
        .await
        .unwrap();
    create_frontend_client(shared_enclave_info(), FRONTEND_SERVICE_ADDR, cred)
        .await
        .unwrap()
}

async fn register_gbdt_function(
    client: &mut TeaclaveFrontendClient<CredentialService>,
) -> ExternalID {
    let fn_input = FunctionInput::new("training_data", "Input traning data file.", false);
    let fn_output = FunctionOutput::new("trained_model", "Output trained model.", false);
    let fn_args = vec![
        FunctionArgument::new("feature_size", "", true),
        FunctionArgument::new("max_depth", "", true),
        FunctionArgument::new("iterations", "", true),
        FunctionArgument::new("shrinkage", "", true),
        FunctionArgument::new("feature_sample_ratio", "", true),
        FunctionArgument::new("data_sample_ratio", "", true),
        FunctionArgument::new("min_leaf_size", "", true),
        FunctionArgument::new("loss", "", true),
        FunctionArgument::new("training_optimization_level", "", true),
    ];

    // Register Function
    let request = RegisterFunctionRequestBuilder::new()
        .name("builtin-gbdt-train")
        .description("Native Gbdt Training Function")
        .arguments(fn_args)
        .inputs(vec![fn_input])
        .outputs(vec![fn_output])
        .build();
    let response = client
        .register_function(request)
        .await
        .unwrap()
        .into_inner();
    log::debug!("Register function: {:?}", response);
    response.function_id.try_into().unwrap()
}

async fn register_input_file(client: &mut TeaclaveFrontendClient<CredentialService>) -> ExternalID {
    let url =
        Url::parse("http://localhost:6789/fixtures/functions/gbdt_training/train.enc").unwrap();
    let crypto = TeaclaveFile128Key::new(&[0; 16]).unwrap();
    let crypto_info = FileCrypto::TeaclaveFile128(crypto);
    let cmac = FileAuthTag::from_hex("860030495909b84864b991865e9ad94f").unwrap();

    let request = RegisterInputFileRequest::new(url, cmac, crypto_info);
    let response = client
        .register_input_file(request)
        .await
        .unwrap()
        .into_inner();
    log::debug!("Register input: {:?}", response);
    response.data_id.try_into().unwrap()
}

async fn register_output_file(
    client: &mut TeaclaveFrontendClient<CredentialService>,
    crypto: impl Into<FileCrypto>,
) -> ExternalID {
    let url =
        Url::parse("http://localhost:6789/fixtures/functions/gbdt_training/e2e_output_model.enc")
            .unwrap();
    let request = RegisterOutputFileRequest::new(url, crypto);
    let response = client
        .register_output_file(request)
        .await
        .unwrap()
        .into_inner();
    log::debug!("Register output: {:?}", response);
    response.data_id.try_into().unwrap()
}

async fn create_gbdt_training_task(
    client: &mut TeaclaveFrontendClient<CredentialService>,
    function_id: &ExternalID,
) -> ExternalID {
    let arguments = FunctionArguments::from_json(serde_json::json!({
        "feature_size": 4,
        "max_depth": 4,
        "iterations": 100,
        "shrinkage": 0.1,
        "feature_sample_ratio": 1.0,
        "data_sample_ratio": 1.0,
        "min_leaf_size": 1,
        "loss": "LAD",
        "training_optimization_level": 2
    }))
    .unwrap();
    let request = CreateTaskRequest::new()
        .executor(Executor::Builtin)
        .function_id(function_id.clone())
        .function_arguments(arguments)
        .inputs_ownership(hashmap!("training_data" => vec![USERNAME]))
        .outputs_ownership(hashmap!("trained_model" => vec![USERNAME]));

    let response = client.create_task(request).await.unwrap().into_inner();
    log::debug!("Create task: {:?}", response);

    response.task_id.try_into().unwrap()
}

async fn assign_data_to_task(
    client: &mut TeaclaveFrontendClient<CredentialService>,
    task_id: &ExternalID,
    training_data_id: ExternalID,
    out_model_id: ExternalID,
) {
    // Assign Data To Task
    let request = AssignDataRequest::new(
        task_id.clone(),
        hashmap!("training_data" => training_data_id),
        hashmap!("trained_model" => out_model_id),
    );
    let response = client.assign_data(request).await.unwrap();

    log::debug!("Assign data: {:?}", response);
}

File: examples/python/builtin_gbdt_train.py
#!/usr/bin/env python3

# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import sys

from teaclave import FunctionInput, FunctionOutput, OwnerList, DataMap
from utils import USER_ID, USER_PASSWORD, connect_authentication_service, connect_frontend_service


class BuiltinGbdtExample:

    def __init__(self, user_id, user_password):
        self.user_id = user_id
        self.user_password = user_password

    def gbdt(self):
        with connect_authentication_service() as client:
            print("[+] login")
            token = client.user_login(self.user_id, self.user_password)

        client = connect_frontend_service()
        metadata = {"id": self.user_id, "token": token}
        client.metadata = metadata

        print("[+] registering function")
        function_id = client.register_function(
            name="builtin-gbdt-train",
            description="Native Gbdt Training Function",
            executor_type="builtin",
            arguments=[
                "feature_size", "max_depth", "iterations", "shrinkage",
                "feature_sample_ratio", "data_sample_ratio", "min_leaf_size",
                "loss", "training_optimization_level"
            ],
            inputs=[
                FunctionInput("training_data", "Input traning data file.")
            ],
            outputs=[FunctionOutput("trained_model", "Output trained model.")])

        print("[+] registering input file")
        url = "http://localhost:6789/fixtures/functions/gbdt_training/train.enc"
        cmac = [
            0x86, 0x00, 0x30, 0x49, 0x59, 0x09, 0xb8, 0x48, 0x64, 0xb9, 0x91,
            0x86, 0x5e, 0x9a, 0xd9, 0x4f
        ]
        schema = "teaclave-file-128"
        key = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
        iv = []
        training_data_id = client.register_input_file(url, schema, key, iv,
                                                      cmac)

        print("[+] registering output file")
        url = "http://localhost:6789/fixtures/functions/gbdt_training/e2e_output_model.enc"
        schema = "teaclave-file-128"
        key = [
            63, 195, 250, 208, 252, 127, 203, 27, 247, 168, 71, 77, 27, 47,
            254, 240
        ]
        iv = []
        output_model_id = client.register_output_file(url, schema, key, iv)

        print("[+] creating task")
        task_id = client.create_task(
            function_id=function_id,
            function_arguments=({
                "feature_size": 4,
                "max_depth": 4,
                "iterations": 100,
                "shrinkage": 0.1,
                "feature_sample_ratio": 1.0,
                "data_sample_ratio": 1.0,
                "min_leaf_size": 1,
                "loss": "LAD",
                "training_optimization_level": 2
            }),
            executor="builtin",
            inputs_ownership=[OwnerList("training_data", [self.user_id])],
            outputs_ownership=[OwnerList("trained_model", [self.user_id])])

        print("[+] assigning data to task")
        client.assign_data_to_task(
            task_id, [DataMap("training_data", training_data_id)],
            [DataMap("trained_model", output_model_id)])

        print("[+] approving task")
        client.approve_task(task_id)

        print("[+] invoking task")
        client.invoke_task(task_id)

        # If you feel like you don't want to wait for the task to finish
        # or something unexpected happens,
        # you may uncomment the following lines to cancel the task
        # time.sleep(3)
        # print("[+] canceling task")
        # client.cancel_task(task_id)

        print("[+] getting result")
        result = client.get_task_result(task_id)
        print("[+] done")
        client.close()

        return bytes(result)


def main():
    example = BuiltinGbdtExample(USER_ID, USER_PASSWORD)
    rt = example.gbdt()

    print("[+] function return: ", rt)


if __name__ == '__main__':
    main()
